### PR TITLE
Add Widget functionality (Address #1100)

### PR DESCRIPTION
**Pull request by SimonJF**
This patch adds the notion of a 'widget' to the MVU library.

A widget is a virtual DOM node which is not affected by the render loop. You can therefore modify it using the usual (real) DOM functionality:

```
import MvuHTML;
import MvuAttrs;
open import Mvu;

typename Message = [| Incr | PopulateWidget |];
typename Model = (counter: Int);

sig view : (Model) ~> MvuHTML.HTML(Message)
fun view(model) {
    open MvuHTML;
    open MvuAttrs;
    var a0 = MvuAttrs.empty;
    var h0 = MvuHTML.empty;

    var widget =
        if (mod(model.counter, 2) == 0) {
                p(style("color:red"),
                    htmlWidget("div", [("id", "widget")]))
            } else {
                p(style("color:blue"),
                    htmlWidget("div", [("id", "widget")]))
            };

    div(a0,
        textNode(intToString(model.counter)) +*
        widget
    ) +*
    button(onClick(fun() { Incr }), textNode("Increment")) +*
    button(onClick(fun() { PopulateWidget }), textNode("Populate Widget"))
}

sig updt : (Message, Model) ~> Model
fun updt(msg, model) {
    switch (msg) {
        case Incr ->
            (counter = model.counter + 1)
        case PopulateWidget ->
            domReplaceChildren(<#>node content</#>, getNodeById("widget"));
            model
    }
}

fun mainPage() {
    var model = (counter = 0);
    runSimple("placeholder", model, view, updt);
    page
        <html><body><div id="placeholder"></div></body></html>
}

serveThis(mainPage)
```

This should address #1100 -- simply create a widget as a placeholder with a known ID, and then populate the widget from an FFI call.

### CODE REVIEW

File: lib/js/vdom.js [hunk 1]
```diff
@@ -541,6 +542,46 @@ const VDom = (function() {
...
+        });
+        return elem;
+      }
+
+      // The innards of a widget are *not* affected by diffing.
+      Widget.prototype.update = function (prev, elem) { }
+
+      // Check to see whether an 'id' has been set explicitly.
+      let id = null;
+      LINKEDLIST.forEach(val[2], function(attr) {
```
Comment (dhil) [inline lib/js/vdom.js:565]: Note for myself: Patch this in #1087.

File: lib/js/vdom.js [hunk 2]
```diff
@@ -541,6 +542,46 @@ const VDom = (function() {
...
+    } else if (lbl == "HTMLWidget") {
+      // Internal widget setup for the virtual-dom library.
+      function Widget(tagName, attrs) {
+        this.type = "Widget";
+        this.tagName = tagName;
+        this.attrs = attrs;
+      }
+      Widget.prototype.init = function () {
+        const elem = document.createElement(this.tagName);
+        LINKEDLIST.forEach(this.attrs, function(attr) {
```
Comment (dhil) [inline lib/js/vdom.js:554]: Patch this in #1087.

File: lib/js/vdom.js [hunk 3]
```diff
@@ -541,6 +542,46 @@ const VDom = (function() {
       return [String(val)];
     } else if (lbl == "HTMLRaw") {
       return [h("span", { innerHTML: String(val) }, [])];
+    } else if (lbl == "HTMLWidget") {
```
Comment (dhil) [inline lib/js/vdom.js:545]: You should probably use `===` here rather than `==` (same two lines above).

File: lib/js/vdom.js [hunk 4]
```diff
@@ -541,6 +542,46 @@ const VDom = (function() {
...
+      LINKEDLIST.forEach(val[2], function(attr) {
+        if (attr[1] == "id") {
+          id = attr[1];
+        }
+      });
+
+      // If the ID has been set, use the ID to persist over renders by
+      // storing in the 'widgets' dictionary.
+      let widget = null;
+      if (id != null) {
```
Comment (dhil) [inline lib/js/vdom.js:574]: Can `id` be `undefined`? If not, then you should probably use `!==` rather than `!=`.

File: lib/js/vdom.js [hunk 5]
```diff
@@ -541,6 +542,46 @@ const VDom = (function() {
...
+          widget = widgets[id];
+        } else {
+          widget = new Widget(val[1], val[2]);
+          widgets[id] = widget;
+        }
+      } else {
+        widget = new Widget(val[1], val[2]);
+      }
+      return [h('div', [widget])];
     } else if (lbl == "HTMLTag") {
```
Comment (dhil) [inline lib/js/vdom.js:585]: `===`

**Review by dhil — APPROVED** [review]
LGTM.

File: lib/js/vdom.js (hunk 3 above)
Comment (dhil) [inline lib/js/vdom.js:545]:
```suggestion
    } else if (lbl === "HTMLWidget") {
```

File: lib/js/vdom.js [hunk 6]
```diff
@@ -541,6 +542,46 @@ const VDom = (function() {
       return [String(val)];
     } else if (lbl == "HTMLRaw") {
```
Comment (dhil) [inline lib/js/vdom.js:543]:
```suggestion
    } else if (lbl === "HTMLRaw") {
```

File: lib/js/vdom.js (hunk 2 above)
Comment (dhil) [inline lib/js/vdom.js:554]:
```suggestion
        _$List.forEach(this.attrs, function(attr) {
```

File: lib/js/vdom.js (hunk 1 above)
Comment (dhil) [inline lib/js/vdom.js:565]:
```suggestion
      _$List.forEach(val[2], function(attr) {
```

File: lib/js/vdom.js [hunk 7]
```diff
@@ -541,6 +542,46 @@ const VDom = (function() {
...
+        return elem;
+      }
+
+      // The innards of a widget are *not* affected by diffing.
+      Widget.prototype.update = function (prev, elem) { }
+
+      // Check to see whether an 'id' has been set explicitly.
+      let id = null;
+      LINKEDLIST.forEach(val[2], function(attr) {
+        if (attr[1] == "id") {
```
Comment (dhil) [inline lib/js/vdom.js:566]:
```suggestion
        if (attr[1] === "id") {
```

File: lib/js/vdom.js (hunk 5 above)
Comment (dhil) [inline lib/js/vdom.js:585]:
```suggestion
    } else if (lbl === "HTMLTag") {
```

**Comment by SimonJF** [comment]
Sorry, this one passed me by -- thanks for the review @dhil. Should be suitable for merging after the CI completes.

**Pull request merged**